### PR TITLE
Apply config updates in global context

### PR DESCRIPTION
**Pull request by amosbird**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Apply configuration updates in global context object. It fixes issues like https://github.com/ClickHouse/ClickHouse/issues/62308.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

---
### Modify your CI run:
**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 8e50c4af4ed2ff1f35d1172ac8e38f8894ff59de with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62944/8e50c4af4ed2ff1f35d1172ac8e38f8894ff59de/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Review by antaljanosbenjamin — APPROVED** [review]
The change looks good, however it is not clear why this only "helps" but not fixes the situation in the mentioned issue?

**Comment by amosbird** [comment]
> The change looks good, however it is not clear why this only "helps" but not fixes the situation in the mentioned issue?

The description was inaccurate. It's actually "fixes".

**Comment by antaljanosbenjamin** [comment]
A lot of errors is tests. There might some issues with our CI or the PR introduces some changes that causes a lot of issues. Let's merge master to start a fresh CI run.

**Review by antaljanosbenjamin — APPROVED** [review]

**Comment by amosbird** [comment]
Test failures are related. The updated configuration does not include the necessary command-line flags. Need to investigate.

**Comment by amosbird** [comment]
@antaljanosbenjamin It seems working.

File: programs/server/Server.cpp [hunk 1]
```diff
@@ -1662,8 +1662,17 @@ try
...
-        [&](ConfigurationPtr config, bool initial_loading)
+        [&, config_file = config().getString("config-file", "config.xml")](ConfigurationPtr config, bool initial_loading)
         {
+            if (!initial_loading)
+            {
+                /// Add back "config-file" key which is absent in the reloaded config.
+                config->setString("config-file", config_file);
+
+                /// Apply config updates in global context.
+                global_context->setConfig(config);
```
Comment (azat) [inline programs/server/Server.cpp:1673]: FWIW @tavplubix shows some concerns about this in the past - https://github.com/ClickHouse/ClickHouse/pull/34574/files#r805753489

Comment (tavplubix) [reply programs/server/Server.cpp:1673]: cc: @amosbird @antaljanosbenjamin 

Also, `setConfig` replaces the `config` object (probably destroying the old one) while `getConfigRef()` returns a reference to this object. Should we revert?

Comment (amosbird) [reply programs/server/Server.cpp:1673]: I'll try to address mentioned issues in #71452

Comment (antaljanosbenjamin) [reply programs/server/Server.cpp:1673]: @tavplubix @azat Thanks for catching these issues, I completely missed these problems and trusted our CI, which I shouldn't have.

**Pull request merged**